modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/defragmentation/DefragmentationFileUtils.java

```diff
@@ -0,0 +1,400 @@
+     * @return File.
+     *
+     * @see DefragmentationFileUtils#defragmentedIndexTmpFile(File)
+     */
+    public static File defragmentedIndexFile(File workDir) {
+        return new File(workDir, DFRG_INDEX_FILE_NAME);
+    }
+
+    /**
+     * Rename temporary index defragmenation file to a finalized one.
+     *
+     * @param workDir Cache group working directory.
+     * @throws IgniteCheckedException If {@link IOException} occurred.
+     *
+     * @see DefragmentationFileUtils#defragmentedIndexTmpFile(File)
+     * @see DefragmentationFileUtils#defragmentedIndexFile(File)
+     */
+    public static void renameTempIndexFile(File workDir) throws IgniteCheckedException {
+        File defragmentedIdxTmpFile = defragmentedIndexTmpFile(workDir);
+        File defragmentedIdxFile = defragmentedIndexFile(workDir);
+
+        try {
+            Files.move(defragmentedIdxTmpFile.toPath(), defragmentedIdxFile.toPath(), ATOMIC_MOVE);
+        }
+        catch (IOException e) {
+            handleIoException(e);
+        }
+    }
+
+    /**
+     * Return file named {@code part-dfrg-%d.bin.tmp} in given folder. It will be used for storing defragmented data
+     * partition during the process.
+     *
+     * @param workDir Cache group working directory.
+     * @param partId Parition index, will be substituted into file name.
+     * @return File.
+     *
+     * @see DefragmentationFileUtils#defragmentedPartFile(File, int)
+     */
+    public static File defragmentedPartTmpFile(File workDir, int partId) {
+        return new File(workDir, String.format(DFRG_PARTITION_TMP_FILE_TEMPLATE, partId));
+    }
+
+    /**
+     * Return file named {@code part-dfrg-%d.bin} in given folder. It will be used for storing defragmented data
+     * partition when the process is over.
+     *
+     * @param workDir Cache group working directory.
+     * @param partId Parition index, will be substituted into file name.
+     * @return File.
+     *
+     * @see DefragmentationFileUtils#defragmentedPartTmpFile(File, int)
+     */
+    public static File defragmentedPartFile(File workDir, int partId) {
+        return new File(workDir, String.format(DFRG_PARTITION_FILE_TEMPLATE, partId));
+    }
+
+    /**
+     * Rename temporary partition defragmenation file to a finalized one.
+     *
+     * @param workDir Cache group working directory.
+     * @param partId Parition index.
+     * @throws IgniteCheckedException If {@link IOException} occurred.
+     *
+     * @see DefragmentationFileUtils#defragmentedPartTmpFile(File, int)
+     * @see DefragmentationFileUtils#defragmentedPartFile(File, int)
+     */
+    public static void renameTempPartitionFile(File workDir, int partId) throws IgniteCheckedException {
+        File defragmentedPartTmpFile = defragmentedPartTmpFile(workDir, partId);
+        File defragmentedPartFile = defragmentedPartFile(workDir, partId);
+
+        assert !defragmentedPartFile.exists() : defragmentedPartFile;
```

Comment:
There may be a race if a user deletes this file manually - should we handle such cases as regular IO exceptions rather than assertions?

Reply:
The assertion says that file "dfrg-part-N.bin" doesn't exist, it can't be there with current code flow.
Anyway, that's an interesting point, I don't expect user to mess with file system during the process.